zephyr/lib/regions_mm.c

```diff
@@ -1,54 +1,574 @@
+ */
+int vmh_free_heap(struct virtual_memory_heap *heap)
+{
+	if (!list_is_empty(&heap->active_allocations))
+		return -ENOTEMPTY;
+
+	for (int i = 0; i < MAX_MEMORY_ALLOCATORS_COUNT; i++) {
+		if (heap->physical_blocks_allocators[i] != 0) {
+			rfree(heap->physical_blocks_allocators[i]
+				  ->bitmap->bundles);
+			rfree(heap->physical_blocks_allocators[i]->bitmap);
+			rfree(heap->physical_blocks_allocators[i]);
+		}
+	}
+	rfree(heap);
+	return 0;
+}
+
+/**
+ * @brief Free ptr allocated on given heap
+ *
+ * Free the ptr allocation. After free action is complete
+ * check if any physical memory block can be freed up as
+ * result.
+ * @param heap Ptr to the heap that is supposed to be freed
+ * @param ptr pointer that should be freed
+ *
+ * @retval 0 on success;
+ * @retval -ENOTEMPTY on heap having active allocations.
+ */
+int vmh_free(struct virtual_memory_heap *heap, void *ptr)
+{
+	struct list_item *iterator;
+	struct vmh_active_allocation *allocation = NULL;
+	int retval;
+
+	list_for_item(iterator, &heap->active_allocations) {
+		if (ptr ==
+		    CONTAINER_OF(iterator, struct vmh_active_allocation, node)
+			->allocation_ptr) {
+
+			allocation = CONTAINER_OF(
+			    iterator, struct vmh_active_allocation, node);
 			break;
-		case MEM_REG_ATTR_OPPORTUNISTIC_MEMORY:
-			vm_heaps[i].memory_caps = SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_HP;
+		}
+	}
+
+	if (allocation == NULL)
+		return -ENOENT;
+	/* Find the allocator that hosts the allocation
+	 * and free when it is found
+	 */
+	for (int i = 0; i < MAX_MEMORY_ALLOCATORS_COUNT; i++) {
+		if (vmh_is_ptr_in_memory_range(
+			ptr, heap->physical_blocks_allocators[i]->buffer,
+			heap->physical_blocks_allocators[i]->num_blocks *
+			    (heap->physical_blocks_allocators[i]->blk_sz_shift
+			     << 1))) {
+			if (heap->span) {
+				retval = sys_mem_blocks_free_contiguous(
+				    heap->physical_blocks_allocators[i], ptr,
+				    allocation->allocation_size);
+			} else {
+				retval = sys_mem_blocks_free(
+				    heap->physical_blocks_allocators[i],
+				    allocation->allocation_size, &ptr);
+			}
 			break;
-		default:
-			return -EINVAL;
 		}
 	}
 
+	list_item_del(&allocation->node);
+
+	/* Scan for changes in physical pages mapping and
+	 * apply unmap if possible
+	 */
+	struct list_item *iterator_phys;
+	struct list_item *tmp;
+	struct list_item *iterator_activ;
+	struct vmh_active_physical_mapping *physical_mAPIng;
+
+	list_for_item_safe(iterator_phys, tmp, &physical_mappings)
+	{
+		bool can_free = true;
+
+		physical_mAPIng = CONTAINER_OF(
+		    iterator_phys, struct vmh_active_physical_mapping, node);
+
+		list_for_item(iterator_activ, &heap->active_allocations) {
+			struct vmh_active_allocation *active_alloc =
+			    CONTAINER_OF(iterator_activ,
+					 struct vmh_active_allocation, node);
+
+			if (vmh_do_ranges_overlap(
+				active_alloc->allocation_ptr,
+				active_alloc->allocation_size,
+				physical_mAPIng->mapping_ptr,
+				physical_mAPIng->mapping_size)) {
```

Comment:
do I understand it correctly that regions a-b and b-c will be reported as overlapping? And that is probably not what you want? And I commented on this before?

Reply:
Fixed in nearest version - I missed that fix in this code version.